diff -Naur c_icap-0.4.2-orig/array.c c_icap-0.4.2/array.c
--- c_icap-0.4.2-orig/array.c	2015-08-27 11:01:50.000000000 +0200
+++ c_icap-0.4.2/array.c	2015-12-15 14:17:25.301679863 +0100
@@ -582,7 +582,7 @@
         if (list->copy_func)
             list->copy_func(it->item, data);
     } else
-        it->item = data;
+        it->item = (void *)data;
     return it;
 }
 
diff -Naur c_icap-0.4.2-orig/include/array.h c_icap-0.4.2/include/array.h
--- c_icap-0.4.2-orig/include/array.h	2015-08-27 11:01:50.000000000 +0200
+++ c_icap-0.4.2/include/array.h	2015-12-15 15:48:16.676975167 +0100
@@ -528,7 +528,7 @@
 #define ci_str_vector_get(vector, i) (i < vector->count ? (const char *)vector->items[i]:  (const char *)NULL)
 #define ci_str_vector_pop(vect)  ((const char *)ci_vector_pop(vect))
 #define ci_str_vector_cast_to_charchar(vector) ((const char **)ci_vector_cast_to_voidvoid(vector))
-#define ci_str_vector_cast_from_charchar(p) (ci_vector_cast_from_voidvoid((void **)p))
+#define ci_str_vector_cast_from_charchar(p) (ci_vector_cast_from_voidvoid((const void **)p))
 
 /**
  * Run the given function for each string vector item
diff -Naur c_icap-0.4.2-orig/mpmt_server.c c_icap-0.4.2/mpmt_server.c
--- c_icap-0.4.2-orig/mpmt_server.c	2015-09-15 11:02:50.000000000 +0200
+++ c_icap-0.4.2/mpmt_server.c	2015-12-15 14:41:03.213712845 +0100
@@ -90,6 +90,7 @@
 #define hard_close_connection(connection)  ci_hard_close(connection->fd)
 #define close_connection(connection) ci_linger_close(connection->fd,MAX_SECS_TO_LINGER)
 void init_commands();
+void commands_exec_scheduled();
 int init_server(char *address, int port, int *family);
 int start_child(int fd);
 void system_shutdown();
@@ -242,7 +243,7 @@
              if (threads_list[i] != NULL) { /* if the i thread is still alive*/
                  if (!threads_list[i]->running) { /*if the i thread is not running any more*/
                      ci_debug_printf(5, "Cancel server %d, thread_id %lu (%d)\n",
-                                     threads_list[i]->srv_id, threads_list[i]->srv_pthread,
+                                     threads_list[i]->srv_id, (long unsigned int)threads_list[i]->srv_pthread,
                                      i);
                      ci_thread_join(threads_list[i]->srv_pthread);
                      release_thread_i(i);
@@ -252,7 +253,7 @@
                      /*The thread is still running, and we have a timeout for waiting 
                        the thread to exit. */
                      if (wait_for_workers <= 2) {
-                         ci_debug_printf(5, "Thread %ld still running near the timeout. Try to kill it\n", threads_list[i]->srv_pthread);
+                         ci_debug_printf(5, "Thread %lu still running near the timeout. Try to kill it\n", (long unsigned int)threads_list[i]->srv_pthread);
                          pthread_kill( threads_list[i]->srv_pthread, SIGTERM);
                      }
                  }
@@ -931,7 +932,7 @@
           commands_exec_scheduled();
      }
 
-     ci_debug_printf(5, "Child :%d going down :%s\n", getpid(),
+     ci_debug_printf(5, "Child :%ld going down :%s\n", (long int)getpid(),
                      child_data->to_be_killed == IMMEDIATELY? 
                      "IMMEDIATELY" : "GRACEFULLY");
      
diff -Naur c_icap-0.4.2-orig/os/unix/proc_utils.c c_icap-0.4.2/os/unix/proc_utils.c
--- c_icap-0.4.2-orig/os/unix/proc_utils.c	2015-08-27 11:01:51.000000000 +0200
+++ c_icap-0.4.2/os/unix/proc_utils.c	2015-12-15 14:42:21.839591521 +0100
@@ -37,7 +37,7 @@
           ci_debug_printf(1, "Cannot open the pid file: %s\n", pidfile);
           return 0;
      }
-     snprintf(strPid, 29, "%d", pid);
+     snprintf(strPid, 29, "%ld", (long int)pid);
      strPid[29] = '\0';
      bytes = write(fd, strPid, strlen(strPid));
      close(fd);
diff -Naur c_icap-0.4.2-orig/registry.c c_icap-0.4.2/registry.c
--- c_icap-0.4.2-orig/registry.c	2015-08-27 11:01:51.000000000 +0200
+++ c_icap-0.4.2/registry.c	2015-12-15 14:21:38.313151380 +0100
@@ -152,5 +152,5 @@
         ci_debug_printf(1, "Registry with id='%d' does not exist!\n", reg_id);
         return 0;
     }
-    return ci_ptr_dyn_array_search(registry, label);
+    return ci_ptr_dyn_array_search((ci_ptr_dyn_array_t *)registry, label);
 }
diff -Naur c_icap-0.4.2-orig/request.c c_icap-0.4.2/request.c
--- c_icap-0.4.2-orig/request.c	2015-08-27 11:01:51.000000000 +0200
+++ c_icap-0.4.2/request.c	2015-12-15 14:26:17.125997147 +0100
@@ -794,7 +794,7 @@
      }
      else if (req->remain_send_block_bytes == CI_EOF) {
 	  if (req->return_code == EC_206 && req->i206_use_original_body >= 0) {
-               def_bytes = sprintf(req->wbuf, "0; use-original-body=%ld\r\n\r\n", 
+               def_bytes = sprintf(req->wbuf, "0; use-original-body=%" PRIu64 "\r\n\r\n", 
 				    req->i206_use_original_body );
 	       req->pstrblock_responce = req->wbuf;
 	       req->remain_send_block_bytes = def_bytes;
diff -Naur c_icap-0.4.2-orig/tests/test_arrays.c c_icap-0.4.2/tests/test_arrays.c
--- c_icap-0.4.2-orig/tests/test_arrays.c	2015-08-27 11:01:51.000000000 +0200
+++ c_icap-0.4.2/tests/test_arrays.c	2015-12-15 15:48:40.788728306 +0100
@@ -119,7 +119,7 @@
     for (j = 1; j < 3; j++) {
         for (i = 1; i< 128; i++) {
             sprintf(value, "value: %d", i);
-            strdata = ci_str_vector_add(vect_str, value);
+            strdata = (char*)ci_str_vector_add(vect_str, value);
             if (!strdata)
                 ci_debug_printf(2, "Can not add: %s\n", value);
         }
@@ -132,9 +132,9 @@
             ci_debug_printf(2, "from charchar value: %s\n", *s);
         }
         ci_str_vector_t *v = ci_str_vector_cast_from_charchar(p);
-        ci_debug_printf(1, "Returned vector max size: %d, itmes %d\n", v->max_size, v->count);
+        ci_debug_printf(1, "Returned vector max size: %d, times %d\n", v->max_size, v->count);
         
-        while((strdata = ci_str_vector_pop(vect_str)) != NULL) {
+        while((strdata = (char*)ci_str_vector_pop(vect_str)) != NULL) {
             ci_debug_printf(2, "Popped value: %s\n", strdata);
         }
     }
diff -Naur c_icap-0.4.2-orig/tests/test_ops.c c_icap-0.4.2/tests/test_ops.c
--- c_icap-0.4.2-orig/tests/test_ops.c	2015-08-27 11:01:51.000000000 +0200
+++ c_icap-0.4.2/tests/test_ops.c	2015-12-15 15:49:35.794622038 +0100
@@ -9,6 +9,7 @@
 #include "mem.h"
 #include "net_io.h"
 
+int mem_init();
 char *str_ip(ci_ip_t *ip)
 {
     char ip_buf[512];
diff -Naur c_icap-0.4.2-orig/tests/test_tables.c c_icap-0.4.2/tests/test_tables.c
--- c_icap-0.4.2-orig/tests/test_tables.c	2015-08-27 11:01:51.000000000 +0200
+++ c_icap-0.4.2/tests/test_tables.c	2015-12-15 14:55:50.818691460 +0100
@@ -88,7 +88,7 @@
     {NULL,NULL,NULL,NULL,NULL}
 };
 
-
+CI_DECLARE_FUNC(int) mem_init();
 int main(int argc,char *argv[]) {
     struct ci_lookup_table *table;
     void *e,*v,**vals;
diff -Naur c_icap-0.4.2-orig/util.c c_icap-0.4.2/util.c
--- c_icap-0.4.2-orig/util.c	2015-08-27 11:01:50.000000000 +0200
+++ c_icap-0.4.2/util.c	2015-12-15 14:24:18.035110038 +0100
@@ -24,6 +24,7 @@
 #include <ctype.h>
 #include <errno.h>
 #include <limits.h>
+#include <stdio.h>
 
 const char *ci_strnstr(const char *s, const char *find, size_t slen)
 {
@@ -149,12 +150,12 @@
 #elif defined(HAVE_STRERROR_R)
     if (strerror_r(error,  buf, buflen) == 0)
         return buf;
-#else
+#endif
+
+/* If nothing defined - no #else to avoid overzealous -Werror=return-type */
     snprintf(buf, buflen, "%d", error);
     buf[buflen - 1] = '\0';
     return buf;
-
-#endif
 }
 
 /*
diff -Naur c_icap-0.4.2-orig/utils/c-icap-client.c c_icap-0.4.2/utils/c-icap-client.c
--- c_icap-0.4.2-orig/utils/c-icap-client.c	2015-08-27 11:01:51.000000000 +0200
+++ c_icap-0.4.2/utils/c-icap-client.c	2015-12-15 14:43:40.053389786 +0100
@@ -360,7 +360,7 @@
 
           if (ret == 206) {
               ci_debug_printf(1, "Partial modification (Allow 206 response): "
-                              "use %ld from the original body data\n", 
+                              "use %" PRIi64 " from the original body data\n", 
                               req->i206_use_original_body);
               copy_data(fd_in, fd_out, req->i206_use_original_body);
 
--- c_icap-0.4.2-orig/configure.ac	2015-09-27 13:04:51.000000000 +0200
+++ c_icap-0.4.2/configure.ac	2015-12-16 11:34:16.000000000 +0100
@@ -349,7 +349,7 @@
       LDFLAGS="$LDFLAGS"
       OLDLIBS=$LIBS
       LIBS="-ldb $OLDLIBS"
-      echo -n "checking for Berkeley DB under standard directories... "
+      AC_MSG_CHECKING([Berkeley DB under standard directories])
       AC_LINK_IFELSE(
                      [
                          #include <db.h>
@@ -375,7 +375,7 @@
       CFLAGS="-I$libdbpath/include $CFLAGS"
       LDFLAGS="-L$libdbpath/lib $LDFLAGS"
       LIBS="-ldb $OLDLIBS"
-      echo -n "checking for Berkeley DB under $libdbpath... "
+      AC_MSG_CHECKING([checking for Berkeley DB under $libdbpath])
       AC_LINK_IFELSE(
                 [
                     #include <db.h>
@@ -411,7 +411,7 @@
      LIBS="-ldb $OLDLIBS"
      for DBVER in 54 53 52 51 50 49 48 47 46 45 44 43 42; do
         if test -d $libdbpath/include/db$DBVER; then
-           echo -n "checking for Berleley DB under $libdbpath/include/db$DBVER and $libdbpath/lib/db$DBVER... "
+           AC_MSG_CHECKING([checking for Berleley DB under $libdbpath/include/db$DBVER and $libdbpath/lib/db$DBVER])
            CFLAGS="-I$libdbpath/include/db$DBVER $OLDCFLAGS"
            LDFLAGS="-L$libdbpath/lib/db$DBVER $OLDLDFLAGS" 
            AC_LINK_IFELSE(
